internal/controller/controllers/preprovisioningimage_controller.go

```diff
@@ -0,0 +1,331 @@
+	// Retrieve InfraEnv
+	infraEnv, err := r.findInfraEnvForPreprovisioningImage(ctx, log, image)
+	if err != nil {
+		log.WithError(err).Error("failed to get corresponding infraEnv")
+		return ctrl.Result{}, err
+	}
+
+	if infraEnv == nil {
+		log.Info("failed to find infraEnv for image")
+		return ctrl.Result{}, nil
+	}
+	if !IronicAgentEnabled(log, infraEnv) {
+		return r.AddIronicAgentToInfraEnv(ctx, log, infraEnv)
+	}
+
+	// The image has been created sooner than the specified cooldown period
+	imageTimePlusCooldown := infraEnv.Status.CreatedTime.Time.Add(InfraEnvImageCooldownPeriod)
+	if imageTimePlusCooldown.After(time.Now()) {
+		log.Info("InfraEnv image is too recent. Requeuing and retrying again soon")
+		return ctrl.Result{Requeue: true, RequeueAfter: time.Until(imageTimePlusCooldown)}, nil
+	}
+
+	if image.Status.ImageUrl == infraEnv.Status.ISODownloadURL {
+		log.Info(ctx, "PreprovisioningImage and InfraEnv images are in sync. Nothing to update.")
+		return ctrl.Result{}, nil
+	}
+	err = r.setImage(image.GetGeneration(), &image.Status, infraEnv.Status.ISODownloadURL, infraEnv.Spec.CpuArchitecture)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	log.Info("updating status")
+	err = r.Status().Update(ctx, image)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	log.Info("PreprovisioningImage Reconcile successfully completed")
+	return ctrl.Result{}, nil
+}
+
+// getConvergedDiscoveryTemplate merge the ironic ignition with the discovery ignition
+func (r *PreprovisioningImageReconciler) getIronicIgnitionConfig(log logrus.FieldLogger, infraEnvInternal common.InfraEnv) (string, error) {
+	config, err := r.IronicIgniotionBuilder.GenerateIronicConfig(r.IronicServiceURL, infraEnvInternal, "")
+	if err != nil {
+		log.WithError(err).Error("failed to generate Ironic ignition config")
+		return "", err
+	}
+	return string(config), err
+}
+
+func (r *PreprovisioningImageReconciler) setImage(generation int64, status *metal3_v1alpha1.PreprovisioningImageStatus, url string, arch string) error {
+	r.Log.Info("Updating PreprovisioningImage ImageUrl to: %s", url)
+	newStatus := status.DeepCopy()
+	newStatus.ImageUrl = url
+	newStatus.Architecture = arch
+	switch r.Format {
+	case models.ImageTypeFullIso:
+		newStatus.Format = metal3_v1alpha1.ImageFormatISO
+	case models.ImageTypeMinimalIso:
+		newStatus.Format = metal3_v1alpha1.ImageFormatInitRD
+	default:
+		return fmt.Errorf("unsupported image format %s", r.Format)
+	}
+	time := metav1.Now()
+	reason := reasonImageWaiting
+	ready := metav1.ConditionFalse
+	message := "waiting for image"
+	if url != "" {
+		reason = reasonImageSuccess
+		ready = metav1.ConditionTrue
+		message = "Generated image"
+	}
+	setImageCondition(generation, newStatus,
+		metal3_v1alpha1.ConditionImageReady, ready,
+		time, reason, message)
+	setImageCondition(generation, newStatus,
+		metal3_v1alpha1.ConditionImageError, metav1.ConditionFalse,
+		time, reason, "")
```

Comment:
The setImage method is called whenever `image.Status.ImageUrl != infraEnv.Status.ISODownloadURL`
I guess it would be better if I'll copy the message and reason from the InfraEnv condition